### PR TITLE
ci: improve macos static linking

### PR DESCRIPTION
**Pull request by kopecs**
In general, we want to statically link to make it easier to distribute
binaries since they won't carry runtime dependencies on various
libraries which would need to be separately installed or managed.
Additionally, this can be a bit fragile depending on which paths the
loader uses, and where dylibs are expected to be (and how possible
differences in the CI build versus other environments).

However, on MacOS, it is a bit difficult to ensure libraries are linked
statically. It is not possible to link all libraries statically.
Notably, libSystem cannot (officially) be statically linked against,
since MacOS does not have a stable syscall ABI, but rather relies on
libSystem being stable. See, e.g.,
- <https://developer.apple.com/forums/thread/706419>
- <https://developer.apple.com/library/archive/qa/qa1118/_index.html>
- `man 1 ld`

As a result, we cannot use a similar strategy to our Alpine build (which
works for Linux, since it is a rare case with a stable syscall ABI, See
<https://unix.stackexchange.com/questions/473137/do-other-unix-like-kernels-have-stable-syscall-abis>).

Previously we relied on a workaround where we deleted the `.dylib` versions
of libraries, leaving only the `.a` versions, thus forcing them to be
linked statically. However, this breaks the build in the case where
later build steps require programs which dynamically linked against
those libraries. Notably, this occurs for PCRE2. See
<https://github.com/semgrep/semgrep/pull/10064>.

One might hope that we could specify the linker to use the `.a` version
without deleting the `.dylib`, but the only way to do this is to disable
autolinking and specify each library individually. Otherwise, on MacOS,
if both are present, the linker will link dynamically regardless.

Here we change the process on MacOS to manually specify each library,
and augment our existing scripts to generate the appropriate list of
flags. This does require marginally more work when adding a library
(since the stubs or library may need to be added to the list), but it
helps to close the gap between what is run locally and in CI, and local
builds will fail in the same way, so debugging is improved.

Test plan: `make test` / existing CI

### CODE REVIEW

**Comment by kopecs** [comment]
@tpetr Do you think
https://github.com/semgrep/semgrep/pull/10122/files#diff-bb37305a4fcd08ae553bde364006478a72895d5608c2617cfecfc631c8025425R28-R34
is still relevant? Or is there an easy bit of cleanup we could do for the same script to run?

**Comment by tpetr** [comment]
@kopecs AFAIK that’s still relevant, but let’s see if the windows job passes in CI (maybe things have changed since I last looked at it). I think you could make things work by explicitly calling bash or sh instead of relying on the shebang for the windows case.

**Comment by kopecs** [comment]
Regarding the Windows stuff, seems like there's an odd CI failure:
```
Error: CreateProcess(): Exec format error
-> required by _build/default/src/main/flags.sexp
-> required by (:include _build/default/src/main/flags.sexp) at
   src/main/dune:4
-> required by _build/default/src/main/Main.exe
-> required by _build/install/default/bin/semgrep-core.exe
```
so just leaving it alone for now. Not sure why given the shell is bash. Maybe there's some reason we can't use `env`?

Anyway, we can still merge this to improve the state of the MacOS stuff and revisit later if needed.

File: src/main/flags.sh
Comment (aryx) [inline src/main/flags.sh:120]: What happens if you remove one?
We get an error in CI?
What happens it tomorrow we add a new language? Will the error be good enough to point to
this file?

**Review by aryx — APPROVED** [review]
Fantastic!

File: src/main/flags.sh
Comment (kopecs) [reply src/main/flags.sh:120]: You would get an error in CI and locally (if you're on MacOS). A representative error would be:
```
/Applications/Xcode.app/Contents/Developer/usr/bin/make minimal-build
dune build _build/install/default/bin/semgrep-core
Generating flags for OS macosx
File "src/main/dune", line 2, characters 8-12:
2 |  (names Main)
            ^^^^
Undefined symbols for architecture arm64:
  "_octs_create_parser_go", referenced from:
      _camlTree_sitter_go__Parse__parse_source_string_328 in tree_sitter_go.a[4](tree_sitter_go__Parse.o)
      _camlTree_sitter_go__Parse__data_begin in tree_sitter_go.a[4](tree_sitter_go__Parse.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
File "caml_startup", line 1:
Error: Error during linking (exit code 1)
make[1]: *** [minimal-build] Error 1
make: *** [core] Error 2
```
(removing go)

Not quite good enough (imo) in terms of pointing to this file, but obvious and reproducible between CI and locally.

Comment (aryx) [reply src/main/flags.sh:120]: since the error mention src/main/dune
you can probably add a comment in it saying "if you get a linking error like XXX, you probably need to modify .../flags.sh  and add the necessary library

File: src/main/flags.sh
Comment (aryx) [inline src/main/flags.sh]: update the comment to not say -cclib but CCLIB,

File: src/main/flags.sh
Comment (aryx) [inline src/main/flags.sh:69]: add a comment saying we just dynamically link on other OS like ubuntu which should be fine; We only
care about statically linux for Alpine because that's the OS we use in our docker image

File: src/main/flags.sh
Comment (kopecs) [reply src/main/flags.sh:120]: Added a printout from Make in 5a250d5b519220912a55ec234fddb427ed3f5441

File: src/main/flags.sh
Comment (kopecs) [reply src/main/flags.sh]: Done in bebd11b58198b473c3bbd14eccb7c718877a1e75

File: src/main/flags.sh
Comment (kopecs) [reply src/main/flags.sh:69]: Done in bebd11b58198b473c3bbd14eccb7c718877a1e75

File: src/main/flags.sh
Comment (aryx) [inline src/main/flags.sh]: typo, Ubuntu

File: Makefile
Comment (aryx) [inline Makefile]: Fancy!

File: Makefile
Comment (ajbt200128) [inline Makefile:421]: nit: can we also add a comment to maybe BREW_DEPS that says something about static linking?

File: Makefile
Comment (ajbt200128) [inline Makefile]: nit: maybe be a lot more verbose in this error message. I.e. "Dune reported a linker error. If you ran into this after adding a new dependency, then this probably means that that dependency relies on a non-OCaml library that needs to be linked. See src/main/flags.sh on how to do that. If you didn't add a new dependency, you may not have every dependency needed for installation. Try running `make dev-setup` and building again". Feel free to tweak or write something else completely, I just think we need to be super verbose here

File: scripts/osx-setup-for-release.sh
Comment (ajbt200128) [inline scripts/osx-setup-for-release.sh:17]: nit: since this is a new file and I'm assuming you're already testing/thinking about this, can you update/remove this comment?

File: scripts/osx-setup-for-release.sh
Comment (ajbt200128) [inline scripts/osx-setup-for-release.sh:21]: same here re: old comments

File: scripts/osx-setup-for-release.sh
Comment (ajbt200128) [inline scripts/osx-setup-for-release.sh:41]: same re: old comments

File: src/main/flags.sh
Comment (ajbt200128) [inline src/main/flags.sh:41]: can we add a short comment at the top here that's like "if you added a dependency that's causing a linking error, here's how to find out what library is needed (opam depext or similar), here's what to update below etc."

Comment (ajbt200128) [reply src/main/flags.sh:41]: also something about what to do if it's a language v opam package

File: src/main/flags.sh
Comment (ajbt200128) [inline src/main/flags.sh:108]: nit: I think the "how to add a dependency" should be higher up. I wouldn't be smart enough to read this entire comment thinking there's a solution at the end, so we shouldn't bury it down here

File: Makefile
Comment (ajbt200128) [inline Makefile:124]: for this PR can you generate one of the error messages and confirm it's readable and easy to debug?

Comment (ajbt200128) [reply Makefile:124]: I saw one in pad's comments. Would be nice if it was top level in the PR comment for future reference. Also would be nice to maybe add a comment here or in flags.sh explaining how to read it.

**Review by ajbt200128 — APPROVED** [review]
Looks good, my comments are all just about adding more comments/being more verbose about everything. I think my main worry is that this isn't idiot proof enough w.r.t. error messages and fixing things. I think it's fine if people have to add a dependency and then think about linking, but since we're shirking a bunch of what dune does normally (autolink), and doing a lot of custom stuff that standard ocaml resources like the dune manual + ocaml discuss don't have a lot of info on, we should make it really easy to understand how to link new things.

File: scripts/osx-setup-for-release.sh
Comment (kopecs) [reply scripts/osx-setup-for-release.sh:17]: This is from a reversion of a prior commit. See [8984ecf](https://github.com/semgrep/semgrep/pull/10122/commits/8984ecf4f2e22708443297bb3d88ccabb4ce0f37)

File: scripts/osx-setup-for-release.sh
Comment (kopecs) [reply scripts/osx-setup-for-release.sh:21]: ditto

File: scripts/osx-setup-for-release.sh
Comment (kopecs) [reply scripts/osx-setup-for-release.sh:41]: ditto

File: src/main/flags.sh
Comment (kopecs) [reply src/main/flags.sh:108]: imo the current format here is better, since it explains why we're doing this, and the part about how to change it is closest to the part you need to change. Per your other comment, I think we could add more info at the top level, so I would rather just have a "Help I have to edit this file and don't know what to do" message there.

File: Makefile
Comment (kopecs) [reply Makefile:421]: Just wrt coupling? Or what are you thinking it would say?

File: Makefile
Comment (kopecs) [reply Makefile]: Sure, that sounds fine.

File: Makefile
Comment (kopecs) [reply Makefile:124]: I mean, it will be as readable as any linker error. I don't think an comment explaining how to read a linker error here would be that useful. What do you think would be unclear about the linking issue itself? I agree with some of the other comments about more instruction on how to change/add to flags.sh.

**Review by tpetr — APPROVED** [review]
LGTM once the oustanding comments are addressed. I have some theories re: the `CreateProcess` error, but I agree that it's not worth holding up this PR for it.

File: Makefile
Comment (kopecs) [reply Makefile:421]: Added one.

File: src/main/flags.sh
Comment (kopecs) [reply src/main/flags.sh:41]: Done.

**Pull request merged**
